### PR TITLE
NOJIRA: Add gradle wrapper, copy over dev deployment tasks

### PR DESCRIPTION
**Pull request by ChristianMurphy**
##### Checklist

-   [x] the [individual contributor license agreement][] is signed
-   [x] commit message follows [commit guidelines][]

##### Description of change

Adds a `build.gradle` and a `gradle wrapper`.
Moves dev deployment tasks over from uPortal.

[individual contributor license agreement]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#individual-contributor-license-agreement
[commit guidelines]: https://github.com/Jasig/uPortal/blob/master/CONTRIBUTING.md#commit
[message properties]: https://github.com/Jasig/uPortal/tree/master/uportal-war/src/main/resources/properties/i18n
[WCAG 2.0 AA]: https://www.w3.org/WAI/WCAG20/quickref/?levels=aaa&technologies=smil%2Cpdf%2Cflash%2Csl

### CODE REVIEW

File: gradle/tasks/clean.gradle [hunk 1]
```diff
@@ -0,0 +1,39 @@
+def props = new Properties()
```
Comment (drewwills) [inline gradle/tasks/clean.gradle]: Static typing is a lot more common in Groovy these days, and may offer advantages in IDEs, etc.

File: gradle/tasks/clean.gradle [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+task loadBuildFile() {
+    doLast {
+        println 'Load the build.properties file'
```
Comment (drewwills) [inline gradle/tasks/clean.gradle]: I'd advocate for interacting with the Gradle `logger` directly:  https://docs.gradle.org/3.3/userguide/logging.html

File: gradle/tasks/clean.gradle [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+        def buildPropsFile = file('build.properties')
+        if (buildPropsFile.exists()) {
+            buildPropsFile.withInputStream {props.load(it)}
+        } else {
+            println('WARNING: build.properties not found')
```
Comment (drewwills) [inline gradle/tasks/clean.gradle]: See above, plus -- should we do something to fail the build here?

File: build.gradle [hunk 4]
```diff
@@ -0,0 +1,2 @@
+apply from: rootProject.file('gradle/tasks/clean.gradle')
```
Comment (drewwills) [inline build.gradle]: There is already a `clean` task in Gradle (and every other build tool).  This file should be renamed to be more specific.

File: gradle/tasks/clean.gradle [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+        }
+    }
+}
+
+task cleanShared(type: Delete, dependsOn: 'loadBuildFile') {
+    group 'Clean Tomcat'
+    description 'Removes ALL shared libraries from the servlet container.'
+    followSymlinks = true
+
+    doFirst {
```
Comment (drewwills) [inline gradle/tasks/clean.gradle]: Evidence suggests that `doLast` is the default/preferred option... unless you need `doFirst`.

File: gradle/tasks/clean.gradle [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+        def serverbase = props.getProperty('server.home')
+        def libs = "${serverbase}/shared/lib"
+        println "Configured shared libraries from servlet container at location:  ${libs}"
```
Comment (drewwills) [inline gradle/tasks/clean.gradle]: see `logger` above.

File: gradle/tasks/clean.gradle [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+        delete libs
+    }
+}
+
+task cleanPortal(type: Delete, dependsOn: 'loadBuildFile') {
+    group 'Clean Tomcat'
+    description 'Removes the deployed uPortal from the servlet container.'
+    followSymlinks = true
+
+    doFirst {
```
Comment (drewwills) [inline gradle/tasks/clean.gradle]: See `doLast` above.

File: gradle/tasks/clean.gradle (hunk 5 above)
Comment (bjagg) [inline gradle/tasks/clean.gradle]: `doFirst` is required in this instance

File: gradle/tasks/clean.gradle (hunk 7 above)
Comment (bjagg) [inline gradle/tasks/clean.gradle]: `doFirst` is required in this instance

File: gradle/tasks/clean.gradle (hunk 1 above)
Comment (ChristianMurphy) [inline gradle/tasks/clean.gradle]: Types are not checked in Groovy, they are more hints.
If we want to do typing I would be more comfortable converting these scripts to Kotlin.

https://blog.gradle.org/kotlin-meets-gradle
https://github.com/gradle/kotlin-dsl
https://docs.gradle.org/current/release-notes.html

File: gradle/tasks/clean.gradle (hunk 5 above)
Comment (ChristianMurphy) [inline gradle/tasks/clean.gradle]: Will do.
For posterity could you link to the reference you found for `doLast` being preferred?

File: build.gradle (hunk 4 above)
Comment (ChristianMurphy) [inline build.gradle]: What would be a preferred name?
Would `clean` be okay if these additional tasks were attached to the `clean` task/phase?

**Comment by drewwills** [comment]
@ChristianMurphy and @stalele -- I'll post some updates to this PR and merge it... I'll work on it this morning

File: gradle/tasks/clean.gradle (hunk 5 above)
Comment (drewwills) [inline gradle/tasks/clean.gradle]: Okay, buy why?  (I'll add a comment.)

File: build.gradle (hunk 4 above)
Comment (drewwills) [inline build.gradle]: Probably cleanShard -- matching the task.

File: gradle/tasks/clean.gradle (hunk 5 above)
Comment (bjagg) [inline gradle/tasks/clean.gradle]: The Delete task needs the variable `delete` to be set to the file/directory to be deleted. This is done in `doFirst`, before it is needed by the parent task.

File: .gitignore [hunk 8]
```diff
@@ -410,7 +410,7 @@ modules.xml
 
 # Project specific ignores
 /build.properties
-/data/
+/hsqldb/
```
Comment (ChristianMurphy) [inline .gitignore]: should hsql files be kept in `build/hsqldb`?

File: README.md [hunk 9]
```diff
@@ -0,0 +1,12 @@
+# uPortal-start
```
Comment (ChristianMurphy) [inline README.md]: :star: Thanks for adding a readme with docs on these tasks! :+1: 
Resolves #2

File: build.gradle [hunk 10]
```diff
@@ -0,0 +1,6 @@
+plugins {
+    id "com.wiredforcode.spawn" version "0.8.2"
```
Comment (ChristianMurphy) [inline build.gradle]: should this plugin import be kept in the `hsql.gradle` file?

File: .gitignore (hunk 8 above)
Comment (drewwills) [inline .gitignore]: I'm thinking not, since a `clean` will delete the whole thing.

File: build.gradle (hunk 10 above)
Comment (drewwills) [inline build.gradle]: I tried that first.  It can't be there -- has to be in a project `build.gradle` file (not one that is imported).

**Pull request merged**